### PR TITLE
Shard 2622

### PR DESCRIPTION
**Pull request by mssabr01**
### **PR Type**
Bug fix, Tests

___

### **Description**
- Fixed edge case handling in `modexp` (0x05) precompile logic.

- Corrected output for zero-length and zero-modulus cases in `modexp`.

- Added comprehensive fuzz and edge case tests for `modexp` precompile.

- Updated ShardeumFlags test to restore check for `disableSmartContractEndpoints`.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>05-modexp.ts</strong><dd><code>Fix edge case handling and output in modexp precompile</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/evm_v2/precompiles/05-modexp.ts

<li>Fixed handling for zero-length and zero-modulus cases.<br> <li> Ensured correct output (empty array) for specific edge cases.<br> <li> Refactored result formatting for consistency.<br> <li> Removed redundant code and improved clarity.

</details>

  </td>
  <td><a href="https://github.com/shardeum/shardeum/pull/584/files#diff-beabac7c1f5c034c0416445ed133f3632f9462501f52c1400b6e9d95635983c5">+17/-16</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>modexpPrecompile.test.ts</strong><dd><code>Add fuzz and edge case tests for modexp precompile</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

test/testCases/modexpPrecompile.test.ts

<li>Added new test suite for MODEXP precompile.<br> <li> Included fuzzer and EIP-198 edge case tests.<br> <li> Tested gas exhaustion and input padding behavior.

</details>

  </td>
  <td><a href="https://github.com/shardeum/shardeum/pull/584/files#diff-1837e615090de5bef9651afb2d5e92529eb17e02b3847c3b2135baead1bb71e8">+59/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>shardeumFlags.test.ts</strong><dd><code>Restore and update ShardeumFlags smart contract flag test</code></dd></summary>
<hr>

test/unit/src/shardeum/shardeumFlags.test.ts

<li>Restored test for <code>disableSmartContractEndpoints</code> flag.<br> <li> Ensured correct flag value in test assertions.

</details>

  </td>
  <td><a href="https://github.com/shardeum/shardeum/pull/584/files#diff-5923ea940501da3a95aa3d918260e012e532c8bbc64df840159f8c00e41d8801">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Additional files</strong></td><td><table>
<tr>
  <td><strong>modexp-testdata.ts</strong></td>
  <td><a href="https://github.com/shardeum/shardeum/pull/584/files#diff-afed4d4ac24a2d2c7a1ca803c7ccc3b6f5de6c45112496ba7b06f85255582690">+2920/-0</a></td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about PR-Agent usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

**Comment by mssabr01** [comment]
do security review

**Review by aniketdivekar — DISMISSED** [review]

File: test/unit/src/shardeum/shardeumFlags.test.ts [hunk 1]
```diff
@@ -271,7 +271,7 @@ describe('ShardeumFlags', () => {
         expect(ShardeumFlags.AppliedTxsMaps).toBe(false)
         expect(ShardeumFlags.SaveEVMTries).toBe(false)
         expect(ShardeumFlags.CheckpointRevertSupport).toBe(true)
-        // Removed test for disableSmartContractEndpoints as it has been replaced by network parameter smartContractSupport
+        expect(ShardeumFlags.disableSmartContractEndpoints).toBe(true)
```
Comment (afostr) [inline test/unit/src/shardeum/shardeumFlags.test.ts]: this flag has been removed on dev.  may need to rebase this to confirm.  I think the right thing is to rebase, then confirm that disableSmartContractEndpoints is gone and remove this test

File: src/evm_v2/precompiles/05-modexp.ts [hunk 2]
```diff
@@ -126,19 +126,13 @@ export function precompile05(opts: PrecompileInput): ExecResult {
     return OOGResult(opts.gasLimit)
   }
 
-  if (bLen === BigInt(0)) {
+  if (bLen === BigInt(0) && mLen === BigInt(0)) {
```
Comment (afostr) [inline src/evm_v2/precompiles/05-modexp.ts:129]: since we are combining our else checks please add the simple matrix of unit tests for 
blen.  zero or non zero
mlen  zero or non zero
and confirm for this matrix of 4 tests that the output is as expected.

File: test/unit/src/shardeum/shardeumFlags.test.ts (hunk 1 above)
Comment (devendra-shardeum) [reply test/unit/src/shardeum/shardeumFlags.test.ts]: changing it back.. thanks for reviewing..

File: src/evm_v2/precompiles/05-modexp.ts (hunk 2 above)
Comment (devendra-shardeum) [reply src/evm_v2/precompiles/05-modexp.ts:129]: The test cases are located in modexp-testdata.ts. It contains a large number of test cases and also covers the matrix.

image

These test cases were taken from the Ethereum monorepo:
https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/evm/test/precompiles/modexp-testdata.ts

You can refer to the full discussion explaining why these changes were needed here:
https://github.com/ethereumjs/ethereumjs-monorepo/issues/3168

This has already been tested, and the results were shared in the following ticket:
https://linear.app/shm/issue/SHARD-2622/correcting-edge-cases-issues-with-modexp-precompile#comment-90c2aad6

@afostr
